Kudu.Core/Jobs/BaseJobRunner.cs

```diff
@@ -72,17 +71,65 @@ protected string WorkingDirectory
 
         protected JobSettings JobSettings { get; set; }
 
-        private static int CalculateHashForJob(string jobBinariesPath)
+        internal static bool JobDirectoryHasChanged(
+            Dictionary<string, FileInfoBase> sourceDirectoryFileMap, 
+            Dictionary<string, FileInfoBase> workingDirectoryFileMap, 
+            Dictionary<string, FileInfoBase> cachedSourceDirectoryFileMap)
         {
-            var updateDatesString = new StringBuilder();
-            DirectoryInfoBase jobBinariesDirectory = FileSystemHelpers.DirectoryInfoFromDirectoryName(jobBinariesPath);
+            // enumerate all source directory files, and compare against the files
+            // in the working directory (i.e. the cached directory)
+            FileInfoBase foundEntry = null;
+            foreach (var entry in sourceDirectoryFileMap)
+            {  
+                if (workingDirectoryFileMap.TryGetValue(entry.Key, out foundEntry))
+                {
+                    if (entry.Value.LastWriteTimeUtc != foundEntry.LastWriteTimeUtc)
+                    {
+                        // source file has changed since we last cached it, or a source
+                        // file has been modified in the working directory.
+                        return true;
+                    }
+                }
+                else
+                {
+                    // A file exists in source that doesn't exist in our working
+                    // directory. This is either because a file was actually added
+                    // to source, or a file that previously existed in source has been
+                    // deleted from the working directory.
```

Comment:
'deleted from the working directory' is hopefully not something that would happen in real world use!